packages/api/src/utils/client/httpClient.ts

```diff
@@ -49,62 +65,183 @@ export type HttpClientModules = {
+        }
+      });
+    }
   }
 
   async json<T>(opts: FetchOpts): Promise<T> {
-    return await this.requestWithBody<T>(opts, (res) => res.json() as Promise<T>);
+    return await this.requestWithBodyWithRetries<T>(opts, (res) => res.json() as Promise<T>);
   }
 
   async request(opts: FetchOpts): Promise<void> {
-    return await this.requestWithBody<void>(opts, async (_res) => void 0);
+    return await this.requestWithBodyWithRetries<void>(opts, async (_res) => void 0);
   }
 
   async arrayBuffer(opts: FetchOpts): Promise<ArrayBuffer> {
-    return await this.requestWithBody<ArrayBuffer>(opts, (res) => res.arrayBuffer());
+    return await this.requestWithBodyWithRetries<ArrayBuffer>(opts, (res) => res.arrayBuffer());
   }
 
-  private async requestWithBody<T>(opts: FetchOpts, getBody: (res: Response) => Promise<T>): Promise<T> {
+  private async requestWithBodyWithRetries<T>(opts: FetchOpts, getBody: (res: Response) => Promise<T>): Promise<T> {
+    // Early return when no fallback URLs are setup
+    if (this.urlsOpts.length === 1) {
+      return this.requestWithBody(this.urlsOpts[0], opts, getBody);
+    }
+
+    let i = 0;
+
+    // Goals:
+    // - if first server is stable and responding do not query fallbacks
+    // - if first server errors, retry that same request on fallbacks
+    // - until first server is shown to be reliable again, contact all servers
+
+    // First loop: retry in sequence, query next URL only after previous errors
+    for (; i < this.urlsOpts.length; i++) {
+      try {
+        return await new Promise<T>((resolve, reject) => {
+          let requestCount = 0;
+          let errorCount = 0;
+
+          // Second loop: query all URLs up to the next healthy at once, racing them.
+          // Score each URL available:
+          // - If url[0] is good, only send to 0
+          // - If url[0] has recently errored, send to both 0, 1, etc until url[0] does not error for some time
+          for (; i < this.urlsOpts.length; i++) {
+            const urlOpts = this.urlsOpts[i];
+            const {baseUrl} = urlOpts;
+            const routeId = opts.routeId ?? DEFAULT_ROUTE_ID;
+
+            if (i > 0) {
+              this.metrics?.requestToFallbacks.inc({routeId});
+              this.logger?.debug("Requesting fallback URL", {routeId, baseUrl, score: this.urlsScore[i]});
+            }
+
+            // eslint-disable-next-line @typescript-eslint/naming-convention
+            const i_ = i; // Keep local copy of i variable to index urlScore after requestWithBody() resolves
+
+            this.requestWithBody(urlOpts, opts, getBody).then(
+              (res) => {
+                this.urlsScore[i_] = Math.min(URL_SCORE_MAX, this.urlsScore[i_] + URL_SCORE_DELTA_SUCCESS);
+                // Resolve immediately on success
+                resolve(res);
+              },
+              (err) => {
+                this.urlsScore[i_] = Math.max(URL_SCORE_MIN, this.urlsScore[i_] - URL_SCORE_DELTA_ERROR);
```

Comment:
should the score be reduced by half, so that the url will have to prove `working` multiple times to come back to max